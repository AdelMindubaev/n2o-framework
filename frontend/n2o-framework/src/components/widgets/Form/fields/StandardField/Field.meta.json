{
  "id": "test1",
  "label": "InputNumber",
  "labelClass": "testLabelClass",
  "help": "testHelp",
  "dependency": [
    {
      "type" : "enabled",
      "on" : ["test2"],
      "expression" : "test2 == null",
      "applyOnInit" : true
    },
    {
      "type" : "required",
      "on" : ["test3"],
      "expression" : "test3 == null",
      "applyOnInit" : true
    },
    {
      "type" : "reRender",
      "on" : ["name","type"]
    }
  ]
}
