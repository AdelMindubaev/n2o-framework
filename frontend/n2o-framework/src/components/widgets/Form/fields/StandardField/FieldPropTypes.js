import PropTypes from 'prop-types'

export const ControlPropTypes = PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
]).isRequired

export const FieldActionsPropTypes = PropTypes.arrayOf(
    PropTypes.shape({
        label: PropTypes.string.isRequired,
    }),
)
