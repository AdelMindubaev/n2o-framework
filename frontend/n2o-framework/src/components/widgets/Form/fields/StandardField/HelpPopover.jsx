import React from 'react';
import PropTypes from 'prop-types';
import UncontrolledPopover from 'reactstrap/lib/UncontrolledPopover';
import PopoverBody from 'reactstrap/lib/PopoverBody';
import { id } from '../../../../../utils/id';

class HelpPopover extends React.Component {
  constructor(props) {
    super(props);

    this.fieldId = id();
  }

  render() {
    const { help, placement, icon } = this.props;
    return (
      <div className={'n2o-popover'}>
        <Button className={'n2o-popover-btn'} id={this.fieldId} color="link">
          <i className={icon} />
        </Button>
        <UncontrolledPopover
          className={'n2o-popover-body'}
          placement={placement}
          target={this.fieldId}
          trigger={'focus'}
        >
          <PopoverBody>
            <div dangerouslySetInnerHTML={{ __html: help }} />
          </PopoverBody>
        </UncontrolledPopover>
      </div>
    );
  }
}

HelpPopover.propTypes = {
  help: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
};

HelpPopover.defaultProps = {
  placement: 'right',
  icon: 'fa fa-question-circle',
};

export default HelpPopover;
