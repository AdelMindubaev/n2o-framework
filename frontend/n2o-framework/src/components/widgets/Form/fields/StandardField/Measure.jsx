import React from 'react'
import PropTypes from 'prop-types'

const measureStyle = { marginLeft: 5 }

/**
 * Компонент-размерность
 * @param {string} value - размерность. Например: км, л, штук...
 * @param {object} props - остальные пропсы
 * @example
 * <Measure value="м"/>
 */
const Measure = ({ value, ...props }) => (value ? (
    <span style={measureStyle} {...props}>
        {value}
    </span>
) : null)

Measure.propTypes = {
    value: PropTypes.string,
}

export default Measure
