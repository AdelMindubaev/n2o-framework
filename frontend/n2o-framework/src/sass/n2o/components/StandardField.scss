%top-label-container{
  display: flex;
  flex-direction: column;
  label {
    min-height: $font-size-base * $line-height-base;
  }
}

.label- {
  &top-right {
    @extend %top-label-container;
    align-items: flex-end;
  }
  &top-left {
    @extend %top-label-container;
    align-items: flex-start;
  }

  &right label,
  &left label {
    display: flex;
    align-items: flex-start;
    margin-bottom: 0;
  }


  &right {
    display: flex;
    flex-direction: row-reverse;
    label {
      margin-left: 5px;
    }
  }
  &left {
    display: flex;
    label {
      margin-right: 5px;
    }
  }
}

.form-group {
  .label-alignment- {
    &left {
      text-align: left;
      justify-content: flex-start;
    }
    &right {
      text-align: right;
      justify-content: flex-end;
    }
  }
}

.n2o-label {
  display: flex;
  & .n2o-popover {
    cursor: pointer;
    margin: 0 5px;

    & button {
      padding: 0;
      background-color: transparent;
      border: none;

      &:focus {
        outline: none;
      }
    }

    &:hover i {
      color: $link-hover-color;
    }
  }
}

.n2o-form-group {
  margin-bottom: 0.5rem;

  & .n2o-validation-message {
    min-height: 1rem;

    font-size: 0.8rem;
    line-height: 1rem;
  }
}