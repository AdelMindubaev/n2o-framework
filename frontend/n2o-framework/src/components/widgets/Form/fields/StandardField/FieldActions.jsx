import React from 'react';
import { FieldActionsPropTypes } from './FieldPropTypes';

/**
 * Компонент, содержащий экшены поля
 * @param  {array} actions - массив экшенов
 */
const FieldActions = ({ actions }) => {
  return (
    <div>
      {actions.map((action, i) => (
        <button key={i}>action.label</button>
      ))}
    </div>
  );
};

FieldActions.propTypes = {
  actions: FieldActionsPropTypes,
};

FieldActions.defaultProps = {
  actions: [],
};

export default FieldActions;
