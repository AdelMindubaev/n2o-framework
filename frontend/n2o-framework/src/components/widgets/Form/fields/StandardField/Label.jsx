import React from 'react';
import PropTypes from 'prop-types';
import BootstrapLabel from 'reactstrap/lib/Label';
import cx from 'classnames';

import HelpPopover from './HelpPopover';

/**
 * Лейбел поля
 * @param {string} id - уникальный индефикатор
 * @param {string|element} value - текст/элемент лейбела
 * @param {boolean} required - поле обязательное(к лейбелу добавляется звездочка) или нет
 * @param {string} className - дополнительный класс стиля
 * @param {string} style - дополнительный dom стиль
 * @param {string} help - подсказка
 * @param {object} props - остальные пропсы
 * @example
 * <Label value='Телефон'/ required={true} />
 */

const Label = ({ id, value, required, className, style, help, ...props }) => {
  const newProps = {
    className: cx('col-form-label', className),
    style: { display: 'inline-block', ...style },
  };

  return React.isValidElement(value) ? (
    <div className={'n2o-field-label'}>
      {React.cloneElement(value, newProps)}
      {required && value ? (
        <span className="n2o-field-label-required">*</span>
      ) : (
        ''
      )}
      {help && <HelpPopover id={id} help={help} />}
    </div>
  ) : (
    <BootstrapLabel className={cx('n2o-field-label', className)} style={style}>
      {value}
      {required && value ? (
        <span className="n2o-field-label-required">*</span>
      ) : null}
      {help && <HelpPopover id={id} help={help} />}
    </BootstrapLabel>
  );
};

Label.propTypes = {
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  required: PropTypes.bool,
  className: PropTypes.string,
  style: PropTypes.object,
};

Label.defaultProps = {
  className: '',
  style: {},
};

export default Label;
