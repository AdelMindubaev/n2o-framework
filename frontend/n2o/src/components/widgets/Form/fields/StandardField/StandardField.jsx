import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import Control from './Control';
import Label from './Label';

import Measure from './Measure';
import Description from './Description';
import FieldActions from './FieldActions';
import InlineSpinner from '../../../../snippets/Spinner/InlineSpinner';

import { FieldActionsPropTypes } from './FieldPropTypes';

/**
 * Компонент - поле формы
 * @reactProps {string} id - уникальный идентификатор поля
 * @reactProps {boolean} visible - отображать / не отображать Поле
 * @reactProps {string} label - лэйбл поля
 * @reactProps {string} labelClass - css-класс для лейбела
 * @reactProps {string} controlClass - css-класс для контрола
 * @reactProps {object} labelStyle- объект стилей для лейбела
 * @reactProps {object} controlStyle - объект стилей для контрола
 * @reactProps {string} className - css-класс для поля
 * @reactProps {boolean} required - обязательное / необязательное поле
 * @reactProps {boolean} disabled - контрол доступен только для чтения / нет
 * @reactProps {boolean} enabled - контрол активирован / нет
 * @reactProps {string|element} control - строка с названием компонента (тем, которое указано в мэпе index.js) или элемент
 * @reactProps {string} description - описание поля (находится под контролом)
 * @reactProps {string} measure - единица измерения, находится после контрола (например, км, кг, л)
 * @reactProps {object} style - объект с css-стилями для поля
 * @reactProps {object} fieldActions - объект для создания экшенов, связанных с полем
 * @reactProps {function} onChange - вызывается при изменении контрола
 * @reactProps {boolean} loading - показывать лоадер(спиннер) или нет
 * @reactProps {boolean} autofocus - есть автофокус на это поле или нет
 * @reactProps {string} validationClass - css-класс валидации(has-error, has-warning или has-success)
 * @reactProps {object} message  - содержит поле text c текстом сообщения(ошибки)
 * @reactProps {string|node} help - подскзка рядом с лейблом
 * @example
 * <Field onChange={this.onChange}
 *             id='DistanceInput'
 *             control='Input'
 *             label="Расстояние"
 *             measure="км"
 *             description="Введите расстояние от пункта А до пункта Б"
 *             style={display: 'inline-block'}/>
 */
class StandardField extends React.Component {
  /**
   * Базовый рендер компонента
   */
  render() {
    const {
      id,
      value,
      visible,
      label,
      control,
      description,
      measure,
      required,
      className,
      labelPosition,
      labelAlignment,
      labelWidth,
      style,
      fieldActions,
      loading,
      autoFocus,
      labelStyle,
      controlStyle,
      labelClass,
      validationClass,
      controlClass,
      onChange,
      onFocus,
      onBlur,
      placeholder,
      touched,
      message,
      help,
      ...props
    } = this.props;
    const flexStyle = { display: 'flex' };
    const validationMap = {
      'is-valid': 'text-success',
      'is-invalid': 'text-danger',
      'has-warning': 'text-warning'
    };
    const styleHelper = { width: '100%' };
    const labelWidthPixels =
      labelWidth === 'default' ? 180 : labelWidth === 'min' ? undefined : labelWidth;
    const extendedLabelStyle = {
      width: labelWidthPixels,
      flex: labelWidthPixels ? 'none' : undefined,
      ...labelStyle
    };
    return (
      visible && (
        <div
          className={cx('n2o-form-group', 'form-group', className, {
            ['label-' + labelPosition]: labelPosition
          })}
          style={style}
        >
          <Label
            id={id}
            value={label}
            style={extendedLabelStyle}
            className={cx(
              labelClass,
              { ['label-alignment-' + labelAlignment]: labelAlignment },
              'n2o-label'
            )}
            required={required}
            help={help}
          />
          <div style={styleHelper}>
            <div style={flexStyle}>
              <Control
                className={cx(controlClass, { [validationClass]: touched })}
                style={controlStyle}
                placeholder={placeholder}
                visible={visible}
                autoFocus={autoFocus}
                value={value}
                onBlur={onBlur}
                onFocus={onFocus}
                onChange={onChange}
                {...control}
                {...props}
              />
              <Measure value={measure} />
              <FieldActions actions={fieldActions} />
              {loading && <InlineSpinner />}
            </div>
            <Description value={description} />
            <div className={cx('n2o-validation-message', validationMap[validationClass])}>
              {message && message.text}
            </div>
          </div>
        </div>
      )
    );
  }
}

StandardField.propTypes = {
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  control: PropTypes.node,
  visible: PropTypes.bool,
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  autoFocus: PropTypes.bool,
  onChange: PropTypes.func,
  description: PropTypes.string,
  measure: PropTypes.string,
  className: PropTypes.string,
  style: PropTypes.object,
  fieldActions: FieldActionsPropTypes,
  valiastionClass: PropTypes.string,
  loading: PropTypes.bool,
  touched: PropTypes.bool,
  labelWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  labelAlignment: PropTypes.oneOf(['left', 'right']),
  labelPosition: PropTypes.oneOf(['top-left', 'top-right', 'left', 'right']),
  message: PropTypes.object,
  help: PropTypes.oneOf(PropTypes.string, PropTypes.node)
};

StandardField.defaultProps = {
  visible: true,
  required: false,
  control: <input />,
  loading: false,
  className: '',
  style: {},
  enabled: true,
  disabled: false,
  onChange: () => {}
};

export default StandardField;
