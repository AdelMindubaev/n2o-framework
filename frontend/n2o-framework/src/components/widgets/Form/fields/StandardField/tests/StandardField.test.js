import React from 'react'
import TestRenderer from 'react-test-renderer'

import StandardField from '../StandardField'

const toolbar = [
    {
        buttons: [
            {
                src: 'StandardButton',
                id: 'update',
                label: 'button 1',
                icon: 'fa fa-trash',
                actionId: 'update',
                validate: true,
                validatedWidgetId: 'create2_main',
                color: 'primary',
                hint: 'some hint',
                size: 'sm',
            },
        ],
    },
]

it.skip('StandardField рендерится корректно', () => {
    const tree = TestRenderer.create(
        <StandardField
            id="myField"
            value="test"
            visible
            label="Мое поле"
            control="InputText"
            description="Введите значение"
            measure="км"
            required
            className="test"
            style={{ display: 'block' }}
            validationState="error"
            loading={false}
            disabled={false}
            enabled
            labelStyle={{ display: 'block' }}
            controlStyle={{ display: 'block' }}
            labelClass="myLabelClass"
            controlClass="myControlClass"
            placeholder="TEST"
            component={() => null}
        />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
})

it.skip('StandardField рендерится корректно c toolbar', () => {
    const tree = TestRenderer.create(
        <StandardField
            id="myField"
            value="test"
            visible
            label="Мое поле"
            control="InputText"
            description="Введите значение"
            measure="км"
            required
            className="test"
            style={{ display: 'block' }}
            validationState="error"
            loading={false}
            disabled={false}
            enabled
            labelStyle={{ display: 'block' }}
            controlStyle={{ display: 'block' }}
            labelClass="myLabelClass"
            controlClass="myControlClass"
            placeholder="TEST"
            component={() => null}
            toolbar={toolbar[0]}
        />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
})

it('StandardField верно приходит props toolbar', () => {
    const tree = TestRenderer.create(
        <StandardField
            id="myField"
            value="test"
            visible
            label="Мое поле"
            control="InputText"
            description="Введите значение"
            measure="км"
            required
            className="test"
            style={{ display: 'block' }}
            validationState="error"
            loading={false}
            disabled={false}
            enabled
            labelStyle={{ display: 'block' }}
            controlStyle={{ display: 'block' }}
            labelClass="myLabelClass"
            controlClass="myControlClass"
            placeholder="TEST"
            component={() => null}
            toolbar={toolbar[0]}
        />,
    ).root
    expect(tree.props.toolbar).toEqual(toolbar[0])
})
