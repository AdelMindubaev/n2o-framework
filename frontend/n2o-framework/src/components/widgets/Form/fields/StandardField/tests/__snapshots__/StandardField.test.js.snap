// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardField рендерится корректно 1`] = `
<div
  className="n2o-form-group form-group test"
  style={
    Object {
      "display": "block",
    }
  }
>
  <label
    className="n2o-field-label myLabelClass n2o-label "
    style={
      Object {
        "display": "block",
        "flex": undefined,
        "width": undefined,
      }
    }
  >
    Мое поле
    <span
      className="n2o-field-label-required"
    >
      *
    </span>
  </label>
  <div
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="control-container"
    >
      <span
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        км
      </span>
      <div />
    </div>
    <div
      style={
        Object {
          "fontSize": "0.8em",
        }
      }
    >
      Введите значение
    </div>
    <div
      className="n2o-validation-message"
    />
  </div>
</div>
`;

exports[`StandardField рендерится корректно c toolbar 1`] = `
<div
  className="n2o-form-group form-group test"
  style={
    Object {
      "display": "block",
    }
  }
>
  <label
    className="n2o-field-label myLabelClass n2o-label "
    style={
      Object {
        "display": "block",
        "flex": undefined,
        "width": undefined,
      }
    }
  >
    Мое поле
    <span
      className="n2o-field-label-required"
    >
      *
    </span>
  </label>
  <div
    style={
      Object {
        "width": "auto",
      }
    }
  >
    <div
      className="control-container control-container_with-toolbar"
    >
      <div
        className="buttons-toolbar n2o-page-actions__form-toolbar btn-toolbar"
        onClick={[Function]}
        role="toolbar"
      >
        <div
          className="btn-group"
          role="group"
        />
      </div>
      <span
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        км
      </span>
      <div />
    </div>
    <div
      style={
        Object {
          "fontSize": "0.8em",
        }
      }
    >
      Введите значение
    </div>
    <div
      className="n2o-validation-message"
    />
  </div>
</div>
`;
